discovery_engine_core/core/src/engine.rs

```diff
@@ -587,9 +587,37 @@ impl Engine {
         if query.trim().is_empty() {
             return Err(Error::InvalidTerm);
         }
+
+        #[cfg(feature = "storage")]
+        if let Ok((search, _)) = self.storage.search().fetch().await {
+            if search.search_by != storage::models::SearchBy::Query || search.search_term != query {
```

Comment:
:grey_question: how about paging, does that need to be checked too?
similar question for `search_by_topic`

Reply:
actually i should fail if there is any search already running (as in the dart code), not just on those conditions. i'm not sure why i wrote it that way, i'll update it accordingly. good catch after all 👍

Reply:
:pushpin: this is currently silently dropping all errors, not just the "no search was open".

Also I personally prefer it if in places where we expect that things are potentially not given we use a `Option` result.

But I can't decide if it applies here.

If this being called with an already open search is a "unexpected situation " (e.g. bug or illegal input) I would say it should be an error. But with it's own error variant, so that we can check on the variant nicely.

But it looks more as if this is somewhat expected in some situations where the query didn't change, in which case I would say it's not falling into the "unexpected" category so I maybe we should change fetch to return a option.

Either way it needs some change in the future.

I used :pushpin:  because I think it's best to do so in a follow up PR so that we can merge pr 0-5 tomorrow to reduce friction (e.g. my PR is based on one of your feature branches as it needs embedding).

Reply:
> this is currently silently dropping all errors, not just the "no search was open".

i took the changes from the following pr which introduces the `Error::NoSearch` variant and changed the return values. now it returns all other errors properly (i think this is also what i had in mind with the original impl).

> But it looks more as if this is somewhat expected in some situations where the query didn't change, in which case I would say it's not falling into the "unexpected" category so I maybe we should change fetch to return a option.

i tend to disagree, even if we create a new query with the very same query term, then it might still return different results, depending on the underlying provider. hence we should treat every new query as distinct queries and not reuse cached data.